clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1491,20 +1495,25 @@ public void setPartition(int partition) {
 
             if (log.isTraceEnabled()) {
                 // Log the message here, because we don't know the partition before that.
-                log.trace("Attempting to append record {} with callback {} to topic {} partition {}", record, userCallback, record.topic(), partition);
+                log.trace("Attempting to append record {} with callback {} to topic {} partition {}", record, userCallback, topic, partition);
             }
+
+            // Reset record to null here so that it doesn't have to be alive as long as the batch is.
+            record = null;
```

Comment:
It's a bit surprising that a method called setPartition resets the record. Maybe we can make the method name clearer. It would also be useful for the comment to state why we no longer need the record after this.

Reply:
The method overrides a callback that is called setPartition to reflect what the caller does with it (it sets partition).  I agree that it's a little non-intuitive to do a state transition here but there doesn't seem to be a better place to do it if we want to preserve the behavior -- we need record until here to do the tracing and we cannot do tracing earlier because we may not know the partition; at the same time, we don't want to keep it longer.

Reply:
We can add another callback method since this is an internal interface, right? This kind of thing leads to a lot of maintenance pain down the line.

Reply:
Thanks for the explanation. Is there a possibility the trace could be done in the caller context? Or is that missing some of the required information?

Reply:
It's missing the partition info. Previously, partition was calculated before doing RecordAccumulator.append so we could do tracing in the doSend, but now the partition may be calculated at the beginning of RecordAccumulator.append, so tracing needs to happen after it's known, but before the actual append proceeds.

Reply:
I find that the code complexity to achieve this trace logging is a bit high. I have some ideas on how to improve it, but we can leave that for latter. A simple suggestion for now would be to change `setPartition` to `onPartitionAssigned` or something like that. This would indicate a general callback that can do anything once the partition is known.

Reply:
Since the only info we need from record is record.partition(), could we keep record.partition() in the instance instead of the whole record? Since record.partition() is much smaller, maybe there is no need to nullify it in setPartition()?

Reply:
Updated to extract all record info in the constructor.